### PR TITLE
Attempting to move color information into the polygons in a way that

### PR DESCRIPTION
**Pull request by madhephaestus**
does not lose color information through CSG ops

https://github.com/NeuronRobotics/JCSG/issues/56

### CODE REVIEW

**Comment by madhephaestus** [comment]
@breiler can you take a look at these changes and see if they fix the issue were having?

**Comment by breiler** [comment]
Thanks, I'll give it try!

**Comment by breiler** [comment]
Given these unit tests it looks like it is doing what it is supposed to, nice work!

A plus for removing the default random color! =)


```java
    @Test
    public void setColor_OnCSGShouldChangeColorsOfAllPolygons() {
        CSG cube = new Cube(10).toCSG()
                .setColor(Color.BLUE);
        assertEquals(Color.BLUE, cube.getColor());

        cube.setColor(Color.RED);

        cube.getPolygons().forEach(polygon -> {
            String polygonColorAsString = getColorAsString(polygon);
            assertEquals("Expected the cube polygons to be another color", RED_COLOR_AS_STRING, polygonColorAsString);
        });
    }

    @Test
    public void setColor_OnUnionedCSGShouldChangeColorsOfAllPolygons() {
        CSG cube1 = new Cube(10).toCSG()
                .setColor(Color.BLUE);

        CSG cube2 = new Cube(10).toCSG()
                .setColor(Color.RED)
                .transformed(new Transform().translate(10, 0, 0));

        CSG union = cube1.union(cube2);
        assertEquals("Expected the new object to to get the default color", CSG.getDefaultColor(), union.getColor());

        union.setColor(Color.BLUE);
        union.getPolygons().forEach(polygon -> {
            String polygonColorAsString = getColorAsString(polygon);
            assertEquals("Expected the cube polygons to be another color", BLUE_COLOR_AS_STRING, polygonColorAsString);
        });
    }
```

**Comment by madhephaestus** [comment]
I am looking into how to pass this information into the JavaFX MeshView. Right now the MeshView flattens the color information because mesh views use a single PhoneMaterial and so can only be a single color... usually. I may be able to make a custom shader material and store the color information in the texture coodrdinates...

**Comment by breiler** [comment]
The PhongMaterial can never have worked/used, right? As it is set on a the complete Mesh and not each polygon? When I debugged the material returned from the mesh it also only had the previously default random color.

I use plain Java and JOGL so I have to do my own hacks to get the colors working:
```java
private void generateBuffers(CSG csg) {
    List<Polygon> polygons = csg.triangulate().getPolygons();

    FloatBuffer vBuffer = Buffers.newDirectFloatBuffer(polygons.size() * 3 * 3);
    FloatBuffer nBuffer = Buffers.newDirectFloatBuffer(polygons.size() * 3 * 3);
    FloatBuffer cBuffer = Buffers.newDirectFloatBuffer(polygons.size() * 3 * 3);

    for (Polygon polygon : polygons) {
        polygon.getPoints().forEach(point -> {
            vBuffer.put((float) point.getX());
            vBuffer.put((float) point.getY());
            vBuffer.put((float) point.getZ());
        });

        nBuffer.put((float) polygon.plane.normal.getX());
        nBuffer.put((float) polygon.plane.normal.getY());
        nBuffer.put((float) polygon.plane.normal.getZ());

        nBuffer.put((float) polygon.plane.normal.getX());
        nBuffer.put((float) polygon.plane.normal.getY());
        nBuffer.put((float) polygon.plane.normal.getZ());

        nBuffer.put((float) polygon.plane.normal.getX());
        nBuffer.put((float) polygon.plane.normal.getY());
        nBuffer.put((float) polygon.plane.normal.getZ());


        Color color = getMatrialColor(polygon);
        cBuffer.put((float)color.getRed());
        cBuffer.put((float)color.getGreen());
        cBuffer.put((float)color.getBlue());

        cBuffer.put((float)color.getRed());
        cBuffer.put((float)color.getGreen());
        cBuffer.put((float)color.getBlue());

        cBuffer.put((float)color.getRed());
        cBuffer.put((float)color.getGreen());
        cBuffer.put((float)color.getBlue());
    }

    this.vertexBuffer = vBuffer;
    this.normalBuffer = nBuffer;
    this.colorBuffer = cBuffer;
}
```

File: src/main/java/eu/mihosoft/vrl/v3d/Polygon.java [hunk 1]
```diff
@@ -728,5 +730,12 @@ public ArrayList<Edge> edges() {
 		return e;
 	}
 
+	public Polygon setColor(Color color) {
+		this.color = color;
+		return this;
+	}
+	public Color getColor() {
```
Comment (madhephaestus) [inline src/main/java/eu/mihosoft/vrl/v3d/Polygon.java:737]: @breiler this should let you get teh color of the specific polygon unaltered by the CSG operationns

Comment (breiler) [reply src/main/java/eu/mihosoft/vrl/v3d/Polygon.java:737]: Sorry, I didn't include that function, that is what I do. The problem I had was setting the color which this PR will fix.

Comment (breiler) [reply src/main/java/eu/mihosoft/vrl/v3d/Polygon.java:737]: I misunderstood what you meant by this, yes this contract change made things a lot more simple. Thanks!

**Pull request merged**